connectors/migrations/20230828_notion_block_parents.ts

```diff
@@ -0,0 +1,132 @@
+// Some Notion resources (pages or databases) have a parent that is a `block`.
+// We don't indent blocks, so we attempt to go up the tree until we find a
+// non-block parent.
+// This migration attempts to backfill the parent of all blocks that have a block parent
+
+import PQueue from "p-queue";
+
+import { getBlockParentMemoized } from "@connectors/connectors/notion/lib/notion_api";
+import { Connector, NotionDatabase, NotionPage } from "@connectors/lib/models";
+import { nango_client } from "@connectors/lib/nango_client";
+import mainLogger from "@connectors/logger/logger";
+
+const logger = mainLogger.child({
+  migration: "20230828_notion_block_parents",
+});
+
+async function main() {
+  const { NANGO_NOTION_CONNECTOR_ID } = process.env;
+
+  if (!NANGO_NOTION_CONNECTOR_ID) {
+    throw new Error("NANGO_NOTION_CONNECTOR_ID not set");
+  }
+
+  const pagesAffected = await NotionPage.findAll({
+    where: {
+      parentType: "block",
+    },
+  });
+  const databasesAffected = await NotionDatabase.findAll({
+    where: {
+      parentType: "block",
+    },
+  });
+
+  const totalResourcesAffected =
+    pagesAffected.length + databasesAffected.length;
+
+  console.log(`Found ${totalResourcesAffected} resources with a block parent`);
+
+  const connectorIds = Array.from(
+    new Set([
+      ...pagesAffected.map((page) => page.connectorId),
+      ...databasesAffected.map((database) => database.connectorId),
+    ])
+  );
+
+  const connectors = await Connector.findAll({
+    where: {
+      id: connectorIds,
+    },
+  });
+
+  const notionAccessTokenByConnectorId: { [key: number]: string } = {};
+  for (const connector of connectors) {
+    if (notionAccessTokenByConnectorId[connector.id]) {
+      continue;
+    }
+    const notionAccessToken = await nango_client().getToken(
+      NANGO_NOTION_CONNECTOR_ID,
+      connector.connectionId
+    );
+    notionAccessTokenByConnectorId[connector.id] = notionAccessToken;
+  }
+
+  const queueByConnectorId: { [key: number]: PQueue } = {};
+  for (const connectorId of connectorIds) {
+    queueByConnectorId[connectorId] = new PQueue({
+      concurrency: 1,
```

Comment:
should this be higher?

Reply:
As per IRL, that's the concurrency per connector (~ per access token), which is kept at 1 to avoid rate-limit issues (and ensure minimal prod sync disruption)

Hopefully fast enough